br/pkg/task/stream.go

```diff
@@ -743,46 +768,48 @@ func RunStreamRestore(
 		return nil
 	}
 	// read data file by given ts.
-	dFiles, mFiles, err := client.ReadStreamDataFiles(ctx, metas, cfg.RestoreTS)
+	dmlFiles, ddlFiles, err := client.ReadStreamDataFiles(ctx, metas, cfg.RestoreTS)
 	if err != nil {
 		return errors.Trace(err)
 	}
 
 	// perform restore meta kv files
-	errChMeta := make(chan error)
-	go func() {
-		rawkvClient, err := newRawkvClient(ctx, cfg.PD, cfg.TLS)
-		if err != nil {
-			errChMeta <- err
-			return
-		}
-		defer rawkvClient.Close()
+	schemasReplace, err := client.InitSchemasReplaceForDDL(&fullBackupTables, cfg.TableFilter)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	rawkvClient, err := newRawkvClient(ctx, cfg.PD, cfg.TLS)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	defer rawkvClient.Close()
 
-		err = client.RestoreMetaKVFiles(ctx, rawkvClient, mFiles, &fullBackupTables)
-		errChMeta <- err
-	}()
+	pm := g.StartProgress(ctx, "Restore DDL files", int64(len(ddlFiles)), cfg.LogProgress)
+	if err = withProgress(pm, func(p glue.Progress) error {
+		return client.RestoreMetaKVFiles(ctx, rawkvClient, ddlFiles, schemasReplace, p.Inc)
+	}); err != nil {
+		return errors.Annotate(err, "failed to restore DDL files")
+	}
 
 	// perform restore kv files
-	pd := g.StartProgress(ctx, "Restore KV Files", int64(len(dFiles)), cfg.LogProgress)
-	err = withProgress(pd, func(p glue.Progress) error {
-		return client.RestoreKVFiles(ctx, rewriteRules, dFiles, p.Inc)
-	})
+	rewriteRules, err := initRewriteRules(client, fullBackupTables)
 	if err != nil {
-		return errors.Annotate(err, "failed to restore kv files")
+		return errors.Trace(err)
 	}
+	updateRewriteRules(rewriteRules, schemasReplace)
 
-	if err := <-errChMeta; err != nil {
-		return errors.Trace(err)
+	pd := g.StartProgress(ctx, "Restore DML Files", int64(len(dmlFiles)), cfg.LogProgress)
+	err = withProgress(pd, func(p glue.Progress) error {
+		return client.RestoreKVFiles(ctx, rewriteRules, dmlFiles, p.Inc)
+	})
+	if err != nil {
+		return errors.Annotate(err, "failed to restore DML files")
 	}
 
+	// fix indices
 	pi := g.StartProgress(ctx, "Restore Index", countIndices(fullBackupTables), cfg.LogProgress)
 	err = withProgress(pi, func(p glue.Progress) error {
-		for _, table := range fullBackupTables {
-			if err := client.FixIndicesOfTable(ctx, table.DB.Name.L, table.Info, p.Inc); err != nil {
-				return errors.Annotatef(err, "failed to fix index for table %s.%s", table.DB.Name, table.Info.Name)
-			}
-		}
-		return nil
+		return client.FixIndicesOfTables(ctx, fullBackupTables, p.Inc)
```

Comment:
if we backup full range. can we ignore this step?

Reply:
I guess maybe we need record whether indices are backed up in the backup meta...

Reply:
agree, @joccau you can leave a TODO here, and fix it in another PR.